### DISCUSSION TITLE
Throwing critical errors with Gunicorn without restarting

### DISCUSSION THREAD

**Discussion by supriyo-biswas** [aio-libs/aiohttp#5837] [Q&A]
I'm writing a web app where I need to validate some environment variables which are needed during the course of operation, and exit otherwise. With plain aiohttp, this is done fairly easily with startup handlers:

```python
async def startup_handler(app):
    if os.getenv('SOME_VAR') is None:
        raise Exception('SOME_VAR must be defined for the app to work')
    # other stuff...

app = web.Application()
app.on_startup.append(startup_handler)
```

However, when I try to run the application with gunicorn, it seems that we restart the worker if the startup handler raises an exception:

```
[2021-06-28 14:51:58 +0000] [11658] [INFO] Worker exiting (pid: 11658)
[2021-06-28 14:51:58 +0000] [11659] [INFO] Booting worker with pid: 11659
[2021-06-28 14:51:58 +0000] [11659] [ERROR] Exception in gunicorn worker
Traceback (most recent call last):
  ......
  File "/var/app/server.py", line 30, in startup_handler
    raise Exception('SOME_VAR must be defined for the app to work')
Exception: SOME_VAR must be defined for the app to work
[2021-06-28 14:51:58 +0000] [11689] [INFO] Booting worker with pid: 11689
[2021-06-28 14:51:58 +0000] [11659] [ERROR] Exception in gunicorn worker
```

Is there a way to throw errors from within the startup handler such that gunicorn doesn't restart workers, or another workaround that I can use to achieve this use case?

**Comment by Dreamsorcerer** [comment]
Isn't the point of gunicorn to try and keep a process running? i.e. It would restart if the process stopped.
I suspect this question is better asked to the gunicorn project.

**Reply by supriyo-biswas** [reply]
One could argue if restarting when the startup handler crashes, the `aiohttp.Gunicorn(UVLoop)WebWorker` should exit instead of restarting the app.

But, coming to my question, I guess what I'm asking is "is there a better place to perform the environment variable checks, outside of the startup handler, so that it'll work with gunicorn as well?"

**Reply by Dreamsorcerer** [reply]
I still think you're more likely to get an answer to your question if directed to gunicorn people. I don't see how this is actually related to aiohttp. If I just wrote some custom code without a framework like:
```
do_startup()  # Maybe throw an exception
while True:
     # Handle requests...
```

Wouldn't the same question still apply? i.e. I think this is a general gunicorn question, not anything to do with aiohttp.
I don't know the answer to your question, and nobody else has responded, so I suggest you ask the question to the gunicorn project.
